### DISCUSSION TITLE
The issue of context length limitations in Ollama

### DISCUSSION THREAD

**Discussion by anrgct** [open-webui/open-webui#4246] [General]
# Bug Report

## Description
I raised this issue in the Ollama repository. Please take a look at this issue, https://github.com/ollama/ollama/issues/6026#issuecomment-2254547480, where the author provided two solutions:

1. Specify num_ctx in the modelfile
2. Pass options during the request

I think method 1 has a drawback: when I set a large num_ctx in modelfile, it will occupy a lot of GPU memory even if the model is not in use. Perhaps method 2 is better? We could increase num_ctx during the request, controlling it based on the length of the messages. Alternatively, we could give users a prompt, saying something like "Your request exceeds the context length and has been truncated for processing."

**Comment by justinh-rahb** [comment]
The issue with changing `num_ctx` is that it requires a reload of the model each time it's changed. This can have an even more negative effect on perceived user experience than it would seek to solve.

**Comment by anrgct** [comment]
I think we should try to minimize user settings and automate this process. Modifying the model file is quite troublesome. If we can handle it within the request, we can make this process seamless. Although switching models will cause some delay, it would be more user-friendly this way.

**Comment by scscgit** [comment]
It's definitely not intuitive that some of the chat history gets ignored. Why is there no warning about part of history being lost (or displaying the cutoff part)? Also we can't even set up per-model context lengths? E.g. Llama3.1 supports 128k, but each chat starts with 2048 by default, and there is no way to change the default [you can only set context length per chat]. There should be a setting to make it extend automatically if the chat grows more. If the chat is short, there is no reason to immediately make it use the full length. And sure, you can make ~20k jumps to avoid reloading the model each time, or whatever else to improve the experience...

These are all design flaws, please respond to them accordingly and tell us your implementation plans. As a beginner I had no idea what needs to be configured, but I guess the defaults should never be really used. You should admit to these limitations in the Readme, because it's far from being a reasonable tool.

Though, also after extending it to 128k, my model just stopped responding at all (consuming all CPU). It would be great if there was any status displayed with potential reasons regarding what's happening in the background. Beginners have no idea how to troubleshoot these problems - and maybe I'm even configuring it incorrectly, how would I know? - It would be very user-friendly if the configuration talked about GBs of memory, including a calculation for how much memory the model takes, so that we can set our limits (for example, I use memory=46GB in my .wslconfig for Docker, so anything beyond that would be terrible; and the model will probably take a lot longer to load, while there will be no indicator of this happening as chat just waits for a model, we can't switch to a smaller model to chat in parallel while a larger model loads, and of course, no way to stop this and reduce the context size, as ollama just hangs).
* Is https://github.com/open-webui/open-webui/issues/573 the only open ticket relating to token counts? (Context Length should probably be in the name.) I don't think `enhancement` gives it justice, it's a severe UX problem.

Last but not least, it's possible to input context length outside the limit of what the model supports...

**Reply by justinh-rahb** [reply]
You paying any of the devs salaries? If not, I'd suggest being less _demanding_ with your requests.

**Reply by scscgit** [reply]
@justinh-rahb I'd interested in learning what is the purpose of an issue tracker in your humble opinion. I did my best to compress a wide range of UX concerns that may not be obvious to everyone. And if a project wants to neglect any important features, it would be only fair to inform us about that in some docs. (Nobody is paying me to test their products either, but surely you believe everyone else's time is free.)

**Reply by justinh-rahb** [reply]
@scscgit Thank you for your feedback. However, I must clarify that this is a volunteer-driven project, and all contributors are dedicating their time without compensation. Your tone and demands overlook the reality of our volunteer efforts.

We appreciate your input, but we need you to approach this collaboratively.  Immediate solutions aren’t always possible, and constructive communication will yield better results than lists of demands.

Thank you in advance for your understanding.

**Reply by scscgit** [reply]
@justinh-rahb Some of us may definitely make a PR for this in future, at least if anyone confirms what changes would be accepted, because it's typical for open source projects to reject some PRs. Basically, it would help for places in source code to be pointed out where this could be implemented. As I have no idea about `ollama` implementation, personally I'm wondering if there are any technical blockers, for example an inability to count the number of tokens when converted to each individual target model. The very first feature I'd expect to be a good fit for new contributors is displaying a counter of tokens, which could be hidden by default until it reaches the configured context length.

**Comment by anrgct** [comment]
Ollama is an important AI infrastructure, so let's take a look at how other projects handle it. In [Continue](https://github.com/continuedev/continue), they use the `/api/chat` endpoint for Ollama's API. They pass different num_ctx values in the request options based on various prompt presets, rather than using the default context length of 2048.

**Comment by anrgct** [comment]
@scscgit  It appears there's an option to control num_ctx. https://github.com/ollama/ollama/issues/6026#issuecomment-2277915109

**Reply by scscgit** [reply]
Yeah I probably could've phrased it better - I know about this context length parameter, but there is no way to change its default (per-model etc.), because every new chat starts with the same value and you have to change it manually each time. Also, changing it to a large value can make the model non-responsive, so it definitely sounds wrong to just set up a large value upfront until the chat history actually grows to that size. And when it comes to new users, they may be completely unaware of the need to configure this.

**Reply by anrgct** [reply]
Yeah, the values I set in the settings don't actually work. You have to adjust them every time you start a new conversation. And that slider is way too sensitive - it's super easy to overshoot and end up with a ridiculous value by accident.

**Reply by justinh-rahb** [reply]
> Yeah, the values I set in the settings don't actually work. You have to adjust them every time you start a new conversation. And that slider is way too sensitive - it's super easy to overshoot and end up with a ridiculous value by accident.

They work for me.

**Comment by mianderson2469** [comment]
Something that would be useful is to auto adjust context length based on available memory after a model is loaded

**Comment by dromeuf** [comment]
The discussion of the length of context used is based on Ollama as an example, but the amount of context used in a chat discussion is not calculated without using Ollama either, but also HF transformer with direct download from HuggingFace in Documents system parameters. If you can implement a counter in the discussion assistant, that would be great. Thanks for your great work. David.

**Comment by tryiou** [comment]
having some 'tokens counter' feature , to let user know what kind of "Context Length" is needed for prompt to be processed fully, would help user adjust correctly context for each need, without enforcing automated calculations.

Same about the discussion history, it's quite hard, as simple user, to know what is pruned and lost over time as we add messages, 
having some kind of indicators of the 'memory' usage,
some kind of ratio like : ( history cost + prompt cost ) / Context Length  
Or displaying the 'pruned' part of prompt/history with some different coloring,
that would greatly help user experience. 

SillyTavern offer a optional feature to measure the state of (context need  versus total context length available).
On every past message of the conversation, user can optionally trigger this detail menu and learn more on situation, can be quite helpful.
![alt text](https://i.ibb.co/p63BhxPR/A.png)
